In [1]:
import math

In [2]:
#1.calculate area of a circle:
def area_of_circle(x):
    A = math.pi*math.pow(x,2)
    return A


In [3]:
R = float(input('What is the radius in cm'))
print(f'The area of circle is %.2f' %area_of_circle(R), 'cm2')

The area of circle is 141.03 cm2


In [4]:
#2. say hello to name
name = input('Who are you??')
print('Hello', name ,'Good norning!!')

Hello Harini Good norning!!


In [5]:
#3.calculate volume of sphere
def volume_of_sphere(x):
    A = (4/3)*math.pi*math.pow(x,3)
    return A


In [6]:
R = float(input('What is the radius in cm'))
print(f'The volume of sphere is %.2f' %volume_of_sphere(R), 'cm3')

What is the radius in cm3
The volume of sphere is 113.10 cm3


In [7]:
#4. Copy String in n times
word = input('What is the word you want to multiply')
print('The multiplied word is', word*5)

What is the word you want to multiplyhellp
The multiplied word is hellphellphellphellphellp


In [8]:
#5.Calculate Area of a Triangle
def area_of_triangle(c,d):
    A = (1/2)*c*d
    return A


In [9]:
h = float(input('What is the height of triangle in cm'))
b = float(input('What is the breadth of triangle in cm'))
print('The area of traingle is', area_of_triangle(h,b), 'cm2')

What is the height of triangle in cm7.9
What is the breadth of triangle in cm6.3
The area of traingle is 24.885 cm2


In [10]:
#6.Calculate Interest
P = float(input('What is the principal amount'))
R = float(input('What is the rate of interest'))
T = float(input('What is the time of the investment period'))
print('The simple interest is', P*R*T)


What is the principal amount10000
What is the rate of interest1.8
What is the time of the investment period5
The simple interest is 90000.0


In [11]:
#8.feet to centimeter converter
F = float(input('What is the measurement in feet'))
print('The centimeter equivalent of feet is', F*30.48, 'cm')

What is the measurement in feet4
The centimeter equivalent of feet is 121.92 cm


In [12]:
#9.BMI Calculator
W = float(input('How heavy are you in kg'))
H = float(input('How tall are you in cm'))
print(f"you're BMI is in metrics %.2f" % ((W/math.pow(H,2))*10000))
#print(f"area of the circle = %.3f" %area)

How heavy are you in kg67
How tall are you in cm177
you're BMI is in metrics 21.39


In [13]:
#10. convert Binary to decimal

bin_num = list(str(input("enter the binary digit ")))
print('The binary array is', bin_num)

counter = 0
dec_num = 0
for i in (reversed(bin_num)):
        dec_num += int(i)*math.pow(2,counter)
        counter+=1
print('Converting the given binary to decimal',dec_num)


enter the binary digit 1101
The binary array is ['1', '1', '0', '1']
Converting the given binary to decimal 13.0


[binary conversion logic](https://stackoverflow.com/questions/47318771/converting-binary-to-decimal-in-python-without-built-in-binary-function)

In [14]:
#11.receive two integral numbers in string form and compute their sum and then print it in console.
A = eval(input('The first number '))
B= eval(input('The second number '))
print('The total of two numbers:',A+B)

The first number 2
The second number 78
The total of two numbers: 80


In [15]:
#12& 13.bitwise operators for any two variable all the bitwise operator that are there
a = eval(input('first number '))
b = eval(input('second number '))
print(f'The bitwise sum is %.1f' %(a&b))
print(f'The bitwise or is %.1f' %(a|b))
print(f'The bitwise xor is %.1f' %(a^b))
print(f'The bitwise not of first number is %.1f and second number is %.1f' %(~a ,~b))
print(f'The left shift of first number to second number is %.1f' %(a<<b))
print(f'The right shift of first number to second number is %.1f' %(a>>b))



first number 2
second number 6
The bitwise sum is 2.0
The bitwise or is 6.0
The bitwise xor is 4.0
The bitwise not of first number is -3.0 and second number is -7.0
The left shift of first number to second number is 128.0
The right shift of first number to second number is 0.0


[bitwise operater](https://www.youtube.com/watch?v=29LZHk2pgVc)

In [16]:
#57.construct the given pattern
n=5
c = '* '
counter = 1

for i in range(0,n):
    print(c*counter)
    counter+=1
    if counter == n:
        print(c*counter)
    if counter >=n:
        for i in range(0,n):
            counter-=1
            print(c*counter)
print('\n')

* 
* * 
* * * 
* * * * 
* * * * * 
* * * * 
* * * 
* * 
* 






In [17]:
#14. str1=" this is a sample daTaa to test for the task" Check if ‘is’ present in str1
str1=" this is a sample daTaa to test for the task" 
position = str1.find('is')
print("The 'is' word is present at the index position" , position)


The 'is' word is present at the index position 3


In [18]:
#15. Check if ‘was’ not present in str1
str1=" this is a sample daTaa to test for the task" 
if (str1.find('was') != -1):
    print("'was' is present in the given string")
else:
    print("'was' is not present in the given string")

'was' is not present in the given string


In [19]:
# 16.& 17 Print str1 in uppercase and as title
str1=" this is a sample daTaa to test for the task"
print('In upper case -',str1.upper())
print('title -',str1.title())
#. 18 Get a new string from str1 ‘DatTaa’ as ‘Data’
str1.replace('daTaa','data',-1)


In upper case -  THIS IS A SAMPLE DATAA TO TEST FOR THE TASK
title -  This Is A Sample Dataa To Test For The Task


' this is a sample data to test for the task'

In [20]:
#20. Display ‘sample’ word from ‘ this is another sample string
import re
str2 = 'this is another sample string'
pattern = re.compile(r"\bs\w*e\b")
print(pattern.search(str2))

<re.Match object; span=(16, 22), match='sample'>


  [display pattern using regex matching](https://www.computerhope.com/issues/ch001721.htm)

In [21]:
#21. Display date, month and year form 21-oct-2020
from datetime import date
today = date.today()
ft_date = today.strftime('%d-%b-%Y')
print("Today's date ",ft_date)

Today's date  20-Mar-2022


In [22]:
#22. arr = “10,20,30,40,50,60,70”, get “50,60”
arr =[10,20,30,40,50,60,70]
print('Using positive indexing ',arr[4:6])
#23 . using negative index get "50 ,60"
print('Using negative indexing ', arr[-3:-1])

Using positive indexing  [50, 60]
Using negative indexing  [50, 60]


In [23]:
#24. From credit_str get ‘8888’ credit_str = "xxxx----xxxx-8888------xxxx"
credit_str = "xxxx----xxxx-8888------xxxx"
num = re.compile('8888')
print(num.search(credit_str))
#25. From credit_str get ‘0’ credit_str = "1234-5678-9878-0434"
credit_str = "1234-5678-9878-0434"
print(credit_str[-4])


<re.Match object; span=(13, 17), match='8888'>
0


In [24]:
#26. convert a integer into a string and print it in console.
a=5
print('It is an integer' , type(a))
print('converting it to string', type(str(a)))

It is an integer <class 'int'>
converting it to string <class 'str'>


In [25]:
#27. take two strings as input and concatenate them and then print it
str1 =" Harry potter"
str2 = " and the goblet of fire"
new_str = str1 + str2  #concatenate
print(new_str)

 Harry potter and the goblet of fire


In [26]:
#28. convert the given string into list given_string = " Apples is differnt from Apple Inc"
given_string = " Apples is differnt from Apple Inc"
print(list(given_string.split(" ")))

['', 'Apples', 'is', 'differnt', 'from', 'Apple', 'Inc']


In [27]:
#29. print type of a and get the 1 element a = {(1,2):1,(2,3):2}
a = {(1,2):1,(2,3):2}
print(type(a))
print(a[(1,2)])

<class 'dict'>
1


In [28]:
#30. write a program to get list as 4,3,2 from the list below a=1,2,3,4,5
a=[1,2,3,4,5]
b = a[1:4]
c = b.reverse()
print(b)


[4, 3, 2]


In [29]:
#31. write a program to get odd number only using a slicing a=1,2,3,4,5,6,7,8,9
a=[1,2,3,4,5,6,7,8,9]
b = a[0:-1:2]
print(b)


[1, 3, 5, 7]


In [30]:
#32. create a tuple with (1,2) 5 times
a =(1,2)*5
print(a)

(1, 2, 1, 2, 1, 2, 1, 2, 1, 2)


In [31]:
#33. from given list create separate lists of strings and numbers.use list comprehension gadgets = "Mobile", "Laptop", 100, "Camera", 310.28, "Speakers", 27.00, "Television", 1000, "Laptop Case", "Camera Lens"
gadgets =[ "Mobile", "Laptop", 100, "Camera", 310.28, "Speakers", 27.00, "Television", 1000, "Laptop Case", "Camera Lens"]
only_strings = [x for x in gadgets if type(x) == str]
only_number = [x for x in gadgets if type(x) == int or type(x)==float]
print(only_number)
print(only_strings)

[100, 310.28, 27.0, 1000]
['Mobile', 'Laptop', 'Camera', 'Speakers', 'Television', 'Laptop Case', 'Camera Lens']


[list comprehension](https://www.w3schools.com/python/python_lists_comprehension.asp)

In [32]:
#34. An array is given from the array find the sum of diagonal using list comprehension a_2d=[1, 2, 3, 4, 5, 6, 7, 8, 9]
import numpy as np
a_2d=[[1,2,3],[4,5,6],[7,8,9]]
a_2d_matrix = np.array(a_2d)
print(a_2d_matrix)
dia_ele = [a_2d_matrix[i][i] for i in range(len(a_2d_matrix))]
print('The sum of the diagonal elements in the above matrix is: ',sum(dia_ele))

[[1 2 3]
 [4 5 6]
 [7 8 9]]
The sum of the diagonal elements in the above matrix is:  15


In [33]:
#35. convert number into tuple of words using comprehension num=8976543201
num_word = {0:'zero',1:'one',2:'two',3:'three',4:'four',5:'five',6:'six',7:'seven',8:'eight',9:'nine'}
num =[8,9,7,6,5,4,3,2,0,1]
words = [num_word[n] for n in num]
print(words)

['eight', 'nine', 'seven', 'six', 'five', 'four', 'three', 'two', 'zero', 'one']


In [34]:
#36. print the numbers which are divisible by 2 but are not a multiple of 5 numbers = range(1, 10)
a = np.linspace(1,10,dtype=int,num=10)
print('The given matrix is: ',a)
b = [x for x in a if x%2==0 and x%5!=0]
print('The required matrix is: ',b)

The given matrix is:  [ 1  2  3  4  5  6  7  8  9 10]
The required matrix is:  [2, 4, 6, 8]


In [35]:
#37. print the sum of tuple : use comprehension init_tuple = [(0, 1), (1, 2), (2, 3)]
init_tuple =[(0,1),(1,2),(2,3)]
sum_tuple = [x+y for (x,y) in init_tuple]
print(sum_tuple)

[1, 3, 5]


In [36]:
#38. delete the last coumn from the given 4*4 matrix arr = [[1, 2, 3, 4], [4, 5, 6, 7], [8, 9, 10, 11], [12, 13, 14, 15]]
arr = [[1, 2, 3, 4], [4, 5, 6, 7], [8, 9, 10, 11], [12, 13, 14, 15]]
pop_ele = [x.pop() for x in arr]
arr_mat =np.array(arr)
print(arr_mat)

[[ 1  2  3]
 [ 4  5  6]
 [ 8  9 10]
 [12 13 14]]


In [37]:
#39. get the list of value of dict in sorted way dict = {'c': 97, 'a': 96, 'b': 98}
dict1 = {'c': 97, 'a': 96, 'b': 98}
[(key, value) for (key, value) in sorted(dict1.items(), key=lambda x: x[1])]

[('a', 96), ('c', 97), ('b', 98)]

[sorting values in dictionary](https://careerkarma.com/blog/python-sort-a-dictionary-by-value/#:~:text=To%20sort%20a%20dictionary%20by%20value%20in%20Python%20you%20can,Dictionaries%20are%20unordered%20data%20structures.)

In [38]:
#40. Creating a Dictionary with Integer Keys and print it
mydict ={1:'A',2:'B',3:'C',4:'D',5:'E'}
print(mydict[3],mydict[5],mydict[1])

C E A


In [39]:
#41. Creating a Dictionary with Mixed keys
#46. Add a Nested Key value to nested dictionary Dictionary
mydict2 = {1:'A','two':'AB','mylist':[2,5,3,4],'x':{0:'zero',5:'five'}}
print(mydict2['mylist'][2])
print(mydict2['x'][5])

3
five


In [40]:
#42. Creating an empty Dictionary
mydict3 = {}
print(type(mydict3))

<class 'dict'>


In [41]:
#43 & 44. Creating a Dictionary with dict() method
mydict4 = dict(name = 'Harini', country =' Finland', adress = '1298 main road')
print(mydict4)

{'name': 'Harini', 'country': ' Finland', 'adress': '1298 main road'}


In [42]:
#45. Creating a Nested Dictionary as shown in the below image {1: 'python', 2: 'For', 3: {'A': 'Welcome', 'B': 'To', 'C': 'Python', 'D': 'class'}}
mydict5 = {1: 'python', 2: 'For', 3: {'A': 'Welcome', 'B': 'To', 'C': 'Python', 'D': 'class'}}
print(mydict5)

{1: 'python', 2: 'For', 3: {'A': 'Welcome', 'B': 'To', 'C': 'Python', 'D': 'class'}}


In [43]:
#47. create a dictionary with id:num and value: square for range (0,6) using dict comprehension
a= np.linspace(0,6,num=6,dtype=int)
print((a))
sq_dict = dict([(key,key**2) for key in a])
print(sq_dict)

[0 1 2 3 4 6]
{0: 0, 1: 1, 2: 4, 3: 9, 4: 16, 6: 36}


[dict comprehension](https://stackoverflow.com/questions/1747817/create-a-dictionary-with-list-comprehension)

In [44]:
#48. create a dictionary using dict comprehension to get the odd square only
a= np.linspace(0,6,num=6,dtype=int)
odd_dict=dict([(key,key**2) for key in a if key%2!=0])
print(odd_dict)

{1: 1, 3: 9}


In [45]:
#49. from a given dict {1: 'a', 2: 'b', 3: 'c'} change the key into item and viceversa
old_dict = {1: 'a', 2: 'b', 3: 'c'}
new_dict = dict([(value, key) for key, value in old_dict.items()])
print('old dict ',old_dict)
print('new dict ', new_dict)

old dict  {1: 'a', 2: 'b', 3: 'c'}
new dict  {'a': 1, 'b': 2, 'c': 3}


In [46]:
#50. get a hello from the given dict 

d = {'k1':[{'nest_key':['this is deep','hello']}]}
print('from the dict d :',d['k1'][0]['nest_key'][1])

d1 = {'k1':[1,2,{'k2':['this is tricky',{'tough':[1,2,'hello']}]}]}
print('from dict d1 :', d1['k1'][2]['k2'][1]['tough'][2])

from the dict d : hello
from dict d1 : hello


In [47]:
#51. Divisibility Check of user given 2 input of numbers
number1 = int(input('Type the first number'))
number2 = int(input('Type the second number'))
if (number1%number2 == 0):
    print(number1, ' is divisible by ',number2, 'and the quotient is', int(number1/number2))
    
else:
    print(number1, ' is not divisible by ',number2)

Type the first number56
Type the second number7
56  is divisible by  7 and the quotient is 8


In [48]:
#52. Check if number is even or odd
no1 = int(input('Type the number'))
if (no1%2==0):
    print(no1,'is an even number')
else:
    print('odd number')

Type the number4
4 is an even number


In [49]:
#53. Check whether Entered character is Vowel or not
vowels = ['a','e','i','o','u']
user_char = str(input('Enter the character :'))
if user_char in vowels:
    print('It is a vowel')
else:
    print('It is a consonaunt')

Enter the character :u
It is a vowel


In [50]:
#54. calculate Sum of n positive Integer example: Sum of n positive Integer till 2 is 3
a=int(input('enter the number of positive integers '))
b = [x for x in range(1,a+1)]
print('Sum of n positive Integer till ',a,' is ',sum(b))

enter the number of positive integers 8
Sum of n positive Integer till  8  is  36


In [51]:
#55. calculate the Digit Sum of a number (eg: 123 print 1+2+3=6)
a= list(str(input('Enter a number :')))
b = [int(x) for x in a]
print('The digit sum of the number is :', sum(b))

Enter a number :865
The digit sum of the number is : 19


In [52]:
#56. Count Numbers, Alphabets, and Special Character from user input string
user = input('Enter the string :')
numbers=0
alphabets=0
spl_char =0
for i in range(len(user)):
    if user[i].isdigit():
        numbers+=1
    elif user[i].isalpha():
        alphabets+=1
    else:
        spl_char+=1
print('The entered string is: ',user)
print('The number of digits is: ',numbers)
print('The number of alphabets is: ',alphabets)
print('The number of special characters is: ',spl_char)

Enter the string :4th left cross street *165
The entered string is:  4th left cross street *165
The number of digits is:  4
The number of alphabets is:  17
The number of special characters is:  5


In [53]:
#58. Count number of characters in a string eg : {'h': 1, 'e': 1, 'l': 2, 'o': 1}
input_dict = {'h': 1, 'e': 1, 'l': 2, 'o': 1}
len(input_dict.keys())

4

In [54]:
#59. write a program to compute 5/0 using try except
try:
    print(5/0)
except:
    print('Divide by zero error occured')

Divide by zero error occured


In [55]:
#60. Handle the exception thrown by the code below by using try and except blocks. li=["hello","class"] for i in li: print (i**2)
li =['hello','class']
try:
    for i in range(len(li)):
        print(li[i]**2)
except:
    print('Cannot perform squares with a string')

Cannot perform squares with a string


In [56]:
#61. Write a function that asks for an integer and prints the square of it. Use a while loop with a try,except, else block to account for incorrect inputs.
#example: Input an integer: null An error occurred! Please try again! Input an integer: 2 Thank you, you number squared is: 4
def squares_of_num(x):
        try:
            return x**2
        except:
            print('An error occurred! Please try again!')
            
    
try:
    a = float(input("Input an integer: "))
    print("Thank you, you're number squared is: ", squares_of_num(a))
            
    
except:
    print('An error occurred! Please try again!')
    



            

Input an integer: 7
Thank you, you're number squared is:  49.0


In [57]:
#62. Accept the String and print the words composed of digits only
sentence = input('Enter a string')
digit =[]
for i in range(len(sentence)):
    if (sentence[i].isdigit()):
        digit.append(sentence[i])

if digit!=0:
    print('The digits are: ',digit)
else:
    print('There are no digits in the string')
        

Enter a stringapple 7yellow
The digits are:  ['7']


In [58]:
#63. find the parity outlier from the given sequence use list comprehension seq=1,2,3,5,7
seq=[1,2,3,5,7]

even=[x for x in seq if x%2==0]
odd =[x for x in seq if x%2!=0]

if (len(even)<len(odd)):
    print('The parity outlier is :',even[0])
else:
    print('The parity outlier is :',odd[0])

The parity outlier is : 2


In [59]:
#64. 
def complement(dna):
    '''a function to find the complement of the DNA sequence'''
    definition = {'A':'T','T':'A','C':'G','G':'C'}
    return ''.join([definition[i] for i in dna[::-1]])

print(complement("ATTGC"))

#dictionary is used within the list comprehension. 


GCAAT


In [60]:
complement.__doc__

'a function to find the complement of the DNA sequence'

In [61]:
#65. Write a program that inputs a number from the user and displays the factorial of that number using while loop.
user_num = int(input("Enter a number to find the factorial"))
counter=user_num
factorial =1
while counter>0:
    factorial=factorial*counter
    counter-=1
    
print('The factorial of ', user_num, 'is', factorial)
    

Enter a number to find the factorial6
The factorial of  6 is 720


In [62]:
#66. Ask the user for a number,to pick the color from the list print the color. color= Blue, Green, Orange, Red
try:
    color = ['Blue','Green','Orange','Red']
    response = int(input('Enter a number between 1 and 4 :'))
    print('Your chosen color is :',color[response-1])
except:
    print('Enter a valid input between 1 and 4')

Enter a number between 1 and 4 :3
Your chosen color is : Orange


In [63]:
#67. Write a program that computes the value of a+aa+aaa+aaaa with a given digit as the value of a.
a = int(input("Enter a number"))
steps = int(input("Enter the number of steps"))
result=0
counter = 1
while counter<=steps:
    result = result + a**counter
    counter+=1

print("The outcome is :",result)

Enter a number5
Enter the number of steps7
The outcome is : 97655


In [69]:
#68. write a program that lets user to input the value seperate by "," and create a list from it using list comprehension
#count = int(input('How many values in the list'))


list=[]
user_value=[]
user_value = input('Enter the values')
list=''.join([x for x in user_value])
print('Entered list is : ',list)

Enter the values3,5,6,3,7
Entered list is :  3,5,6,3,7


In [65]:
#69. Write a program that calculates and prints the value according to the given formula: Q = Square root of (2 C D)/H given, c=50 h=30 D is the variable whose values should be input to your program in a comma-separated sequence. Example 100,150,180 The output : 18,22,24

d =[int(x) for x in input('Enter the values separated by comma :').split(',')]
#print(type(d))

result=[]

for i in range(len(d)):
    Q = math.sqrt((2*50*d[i])/30)
    result.append(int(Q))

print(result)

Enter the values separated by comma :100,150,180
[18, 22, 24]


In [70]:
#70. Write a program that accepts a sequence of whitespace separated words as input and prints the words after removing all duplicate words and sorting them alphanumerically.

user_strings = [str(x) for x in input("Enter the words separated by space: ").split()]
final_string = set(user_strings)
print(sorted(final_string))

Enter the words separated by space: apple pie pudding cream pastry pie cream
['apple', 'cream', 'pastry', 'pie', 'pudding']


In [72]:
#71. Write a program which accepts a sequence of comma separated 4 digit binary numbers and then check whether they are divisible by 5 or not. print the number Example: 0100,0011,1010,1001 Then the output should be: 1010
bin_input =[int(x,2) for x in input("enter the binary numbers separated by comma: ").split(',')]
div_by_5 = [int(x) for x in bin_input if x%5==0]
output = [bin(x)[2:] for x in div_by_5]
print('The binary number divisible by 5 is : ',output)

enter the binary numbers separated by comma: 101,100,110,001
The binary number divisible by 5 is :  ['101']


In [73]:
#72. By using list comprehension, please write a program to print the list after removing the even index numbers li=range(10)
li = [1,2,3,4,5,6,7,8,9,10]
updated_li =[x for x in li if li.index(x)%2!=0]
print(updated_li)

[2, 4, 6, 8, 10]


In [74]:
#73. write a function to get the sum of two num(use lambda)
c = lambda a,b:a+b
print(c(8,4))
print(c(12,6))
print(c(0,7))


12
18
7


In [75]:
#74. write a function to print if leap year or not
def leap_year(x):
    if(x%4==0):
        return 'leap year'
    else:
        return 'not a leap year'
    
year = int(input('Enter the year: '))
print('It is ', leap_year(year))
    
    

Enter the year: 2004
It is  leap year


In [76]:
#75. write a function to count a bit(number of 1) in number (use lambda)

number = input('Enter a binary number')
count = lambda number:number.count('1')
print('the number of bits (1) in the given binary number is: ',count(number))

Enter a binary number110011010
the number of bits (1) in the given binary number is:  5


In [77]:
#76. write a program to get the area of a circle from the radius (use lambda)
radius = float(input('Enter the radius'))
area = lambda radius:math.pi*(radius**2)
print(f'The area of circle is: %.2f'%area(radius),'cm2')

Enter the radius2.4
The area of circle is: 18.10 cm2


In [78]:
#77. write a function to get the factorial of given number use recurssion
def recursive(num):
    if num==1:
        return num
    else:
        return num*recursive(num-1)

num = int(input('Enter a number: '))
print('The factorial is ',recursive(num))

Enter a number: 6
The factorial is  720


In [79]:
#78. write a program to remove a duplicate word in a list or user input sentences.
user_str = [str(x) for x in input("Enter the sentence: ").split()]
final_str = set(user_str)
print(final_str)

Enter the sentence: I am going to espoo espoo in helsinki
{'espoo', 'am', 'going', 'to', 'helsinki', 'in', 'I'}


In [80]:
#79. calculate fibonacci
n1 = 0
n2 = 1
count = 0
num = int(input('Enter the no of terms: '))
if num<=0:
    print('Enter a positive number')
elif (num==1):
    print('the result is ', num)
else:
    print('Fibonnacci series:')
    #print(n1)
    #print(n2)
    while count<num:
            print(n1)
            nth = n1+n2
        #print(nth)
            n1=n2
            n2=nth
            count+=1
        
        

Enter the no of terms: 10
Fibonnacci series:
0
1
1
2
3
5
8
13
21
34


In [81]:
#80. write a function to get the premutation and combination
import itertools
a = [1,5,4,3]
permut = itertools.permutations(a,2)
print('possible permutations are :')
for item in permut:
    print(item)
    
combi = itertools.combinations(a,2)
print('possible combinations are :')
for item in combi:
    print(item)
    


possible permutations are :
(1, 5)
(1, 4)
(1, 3)
(5, 1)
(5, 4)
(5, 3)
(4, 1)
(4, 5)
(4, 3)
(3, 1)
(3, 5)
(3, 4)
possible combinations are :
(1, 5)
(1, 4)
(1, 3)
(5, 4)
(5, 3)
(4, 3)


In [82]:
#81. write a function to draw a diamond shape from a user given number: hint: only odd number can draw a diamond shape
h = int(input('Enter the height of the diamond '))
for x in range(h):
    print(' '*(h-x),'*'*((2*x)+1))
for x in range(h-2,-1,-1):
    print(' '*(h-x),'*'*((2*x)+1))


Enter the height of the diamond 10
           *
          ***
         *****
        *******
       *********
      ***********
     *************
    ***************
   *****************
  *******************
   *****************
    ***************
     *************
      ***********
       *********
        *******
         *****
          ***
           *


[diamond printing in python](https://www.codingem.com/diamond-pattern-in-python-using-for-loop/)

In [83]:
#82. build a simple calculator which can: add subtract multiply divide

try:
    def calculator(a,b,response):
        if response==1:
            return a+b
        elif response==2:
            return a-b
        elif response==3:
            return a*b
        else:
            return a/b

    
    a,b = map(float, input("Enter two values separated by comma: ").split(','))
    response = int(input(" Enter your choice: add -1,sub-2,product-3,division-4: "))
    print("The operation chosen is ", response," and the result is ",calculator(a,b,response))

except ZeroDivisionError:
    print("division by zero. Enter non zero values")
       

Enter two values separated by comma: 4,6
 Enter your choice: add -1,sub-2,product-3,division-4: 4
The operation chosen is  4  and the result is  0.6666666666666666


In [84]:
#83. define a function which can print a given string in a circular rotation


def circular(s):
     for i in range(len(s)):
            rot = s[i:]+s[:i]
            print(rot)
            

instring = input('enter the string to be rotated')
#print(instring)
circular(instring)            
        

enter the string to be rotatedHARINI
HARINI
ARINIH
RINIHA
INIHAR
NIHARI
IHARIN


In [85]:
#84. write a function to calculate the GCD of 2 number

a_div=[]
b_div=[]
def GCD(a,b):
    for i in range(1,a+1):
        if (a%i==0):
            a_div.append(i)
    for i in range(1,b+1):
        if(b%i==0):
            b_div.append(i)
    return max((set(a_div).intersection(set(b_div))))

GCD(6,10)

2

In [86]:
#85. Write a Python function to check whether a string is pangram or not.
import string
a =set()
s = set(string.ascii_lowercase)
def pangram(sent):
    print('Entering function....')
    for i in range(len(sent)):
        if sent[i].lower() in s:
            a.add(sent[i].lower())   
    if len(a) == len(s):
        print('All the alphabets are found in the string')
        return 'String is a pangram'
    else:
        return 'String is not a pangram'
        

        
strng = input('Enter a string')
pangram(strng)


#Pack my box with five dozen liquor jugs
# A quick brown fox jumps over the lazy cat

Enter a stringPack my box with five dozen liquor jugs
Entering function....
All the alphabets are found in the string


'String is a pangram'

In [87]:
#86. write a function which take a list of number and target where target is number . print the index from the list where the sum of the index value is equal to that of the target
def simple(lst,target):
    n = len(lst)
    for first in range(n):
        for scnd in range(first+1,n):
            if (lst[first]+lst[scnd] == target):
                print('The index values whose sum is equal to target',target,' is ',first,scnd)

simple([1,2,5,6,4],10)
simple([2,4,2,6,7],11)
simple([5,3,7,8,4,5],15)
        

The index values whose sum is equal to target 10  is  3 4
The index values whose sum is equal to target 11  is  1 4
The index values whose sum is equal to target 15  is  2 3


In [88]:
#87. write a function to calculate the LCM of a 2 number lcm of 54 and 24 is 216
num1,num2= map(int,input('Enter 2 numbers separated by comma').split(','))

def LCM(n1,n2):
    return ((n1*n2)/GCD(n1,n2))

print(f'The LCM of {num1} and {num2} is' ,LCM(num1,num2))
#GCD(6,10)

Enter 2 numbers separated by comma54,24
The LCM of 54 and 24 is 216.0


In [89]:
#88. write a function that gives the prime factor only

factors=[]
#a=num
def prime_factor(a):
    prime=2
    #a=num
    while(a>1):
        #print(a)
        if(a%prime==0):
            factors.append(prime)
            a=a/prime
            #print (factors)
        else:
            prime+=1
            
    return factors

num = int(input('Enter a number to find its prime factors'))
print(f'The prime factors of {num} is: ',prime_factor(num))
            
        
        
    

Enter a number to find its prime factors60
The prime factors of 60 is:  [2, 2, 3, 5]


In [90]:
#89. write a function that print the first non repeating letter
def firstNonRepeatingChar(str1):
    char_order = []
    counts = {}
    for c in str1:
        if c in counts:
            counts[c] += 1
        else:
            counts[c] = 1
            char_order.append(c)
    for c in char_order:
        if counts[c] == 1:
            return c
    return None

firstNonRepeatingChar('shesellsseashellsontheseashore')

'n'

In [91]:
#90. return sum of array excluding the lowest and highest element
i = [int(x) for x in input('Enter the array').split(',')]
mx = max(i)
mn= min(i)
i.remove(mx)
i.remove(mn)
print(sum(i))

Enter the array2,8,4,89,34,21,4
71


In [92]:
#91. define a function take a string value replace any number below 5 with 0 and any number above 5 with 1
line = input('enter the string')
for i in range(len(line)):
    if line[i].isdigit():
        if int(line[i])>=5:
            line = line.replace(line[i], '1')
        else:
            line = line.replace(line[i],'0')
        
print(line)

enter the string123 musical street 647th cross
000 musical street 101th cross


In [93]:
#92. write a function to get a day of week eg: (30,8,2010)=1
from datetime import date
date.today().strftime('%d,%m,%Y')

'21,03,2022'

In [94]:
#93. write a function to get all the permutation of the string
from itertools import permutations

permu = permutations('harini')
count = 0
for item in (permu):
    print(item)
    count+=1
print('The number of permutations are' ,count)

('h', 'a', 'r', 'i', 'n', 'i')
('h', 'a', 'r', 'i', 'i', 'n')
('h', 'a', 'r', 'n', 'i', 'i')
('h', 'a', 'r', 'n', 'i', 'i')
('h', 'a', 'r', 'i', 'i', 'n')
('h', 'a', 'r', 'i', 'n', 'i')
('h', 'a', 'i', 'r', 'n', 'i')
('h', 'a', 'i', 'r', 'i', 'n')
('h', 'a', 'i', 'n', 'r', 'i')
('h', 'a', 'i', 'n', 'i', 'r')
('h', 'a', 'i', 'i', 'r', 'n')
('h', 'a', 'i', 'i', 'n', 'r')
('h', 'a', 'n', 'r', 'i', 'i')
('h', 'a', 'n', 'r', 'i', 'i')
('h', 'a', 'n', 'i', 'r', 'i')
('h', 'a', 'n', 'i', 'i', 'r')
('h', 'a', 'n', 'i', 'r', 'i')
('h', 'a', 'n', 'i', 'i', 'r')
('h', 'a', 'i', 'r', 'i', 'n')
('h', 'a', 'i', 'r', 'n', 'i')
('h', 'a', 'i', 'i', 'r', 'n')
('h', 'a', 'i', 'i', 'n', 'r')
('h', 'a', 'i', 'n', 'r', 'i')
('h', 'a', 'i', 'n', 'i', 'r')
('h', 'r', 'a', 'i', 'n', 'i')
('h', 'r', 'a', 'i', 'i', 'n')
('h', 'r', 'a', 'n', 'i', 'i')
('h', 'r', 'a', 'n', 'i', 'i')
('h', 'r', 'a', 'i', 'i', 'n')
('h', 'r', 'a', 'i', 'n', 'i')
('h', 'r', 'i', 'a', 'n', 'i')
('h', 'r', 'i', 'a', 'i', 'n')
('h', 'r

In [95]:
#94. write a function to get all the possible subset of a string in a list

from itertools import combinations

instring = 'harini'
count = 0
for i in range(2,len(instring)):
    subset = combinations('harini',i)
    for item in (subset):
        print(item)
        count+=1
print('The number of combinations are' ,count)

('h', 'a')
('h', 'r')
('h', 'i')
('h', 'n')
('h', 'i')
('a', 'r')
('a', 'i')
('a', 'n')
('a', 'i')
('r', 'i')
('r', 'n')
('r', 'i')
('i', 'n')
('i', 'i')
('n', 'i')
('h', 'a', 'r')
('h', 'a', 'i')
('h', 'a', 'n')
('h', 'a', 'i')
('h', 'r', 'i')
('h', 'r', 'n')
('h', 'r', 'i')
('h', 'i', 'n')
('h', 'i', 'i')
('h', 'n', 'i')
('a', 'r', 'i')
('a', 'r', 'n')
('a', 'r', 'i')
('a', 'i', 'n')
('a', 'i', 'i')
('a', 'n', 'i')
('r', 'i', 'n')
('r', 'i', 'i')
('r', 'n', 'i')
('i', 'n', 'i')
('h', 'a', 'r', 'i')
('h', 'a', 'r', 'n')
('h', 'a', 'r', 'i')
('h', 'a', 'i', 'n')
('h', 'a', 'i', 'i')
('h', 'a', 'n', 'i')
('h', 'r', 'i', 'n')
('h', 'r', 'i', 'i')
('h', 'r', 'n', 'i')
('h', 'i', 'n', 'i')
('a', 'r', 'i', 'n')
('a', 'r', 'i', 'i')
('a', 'r', 'n', 'i')
('a', 'i', 'n', 'i')
('r', 'i', 'n', 'i')
('h', 'a', 'r', 'i', 'n')
('h', 'a', 'r', 'i', 'i')
('h', 'a', 'r', 'n', 'i')
('h', 'a', 'i', 'n', 'i')
('h', 'r', 'i', 'n', 'i')
('a', 'r', 'i', 'n', 'i')
The number of combinations are 56


In [96]:
#95. write a function to get the length of the last word only from a string seperated by space ex:('string of words') == 5 ex:('adasd ')=5
def last_word_length():
    instring1 = [str(x) for x in input('Enter the string').split()]
    last_word = instring1.pop()
    print(f'The length of the last word {last_word} is')
    return len(last_word)
    
last_word_length()

Enter the stringThis is the longest python assignment
The length of the last word assignment is


10

In [97]:
#96. write a program for a tower of hanoi puzzle

def TowerOfHanoi(n , source, destination, auxiliary):
    if n==1:
        print ("Move disk 1 from source",source,"to destination",destination)
        return
    TowerOfHanoi(n-1, source, auxiliary, destination)
    print ("Move disk",n,"from source",source,"to destination",destination)
    TowerOfHanoi(n-1, auxiliary, destination, source)
         
n = 4
TowerOfHanoi(n,'A','B','C')

Move disk 1 from source A to destination C
Move disk 2 from source A to destination B
Move disk 1 from source C to destination B
Move disk 3 from source A to destination C
Move disk 1 from source B to destination A
Move disk 2 from source B to destination C
Move disk 1 from source A to destination C
Move disk 4 from source A to destination B
Move disk 1 from source C to destination B
Move disk 2 from source C to destination A
Move disk 1 from source B to destination A
Move disk 3 from source C to destination B
Move disk 1 from source A to destination C
Move disk 2 from source A to destination B
Move disk 1 from source C to destination B


In [98]:
#97. Use a for loop to estimate π from the first 20 terms of the Madhava series:
import math
pi = 0
for k in range(20):
    pi +=(pow(-3, -k) / (2*k+1))
    
pi *= math.sqrt(12)
print(f'pi = %.4f' %pi)

pi = 3.1416


In [99]:
#98. write a function to check if the number is harshad or not A Harshad number is an integer that is divisible by the sum of its digits (for example, 21 is divisible by 2+1=3 and so is a Harshad number).
def harshad_num():
    num = [int(x) for x in input('Enter a number')]
    s = sum(num)
    num1 = one_number(num)
    if (num1%s==0):
        print(f'the number {num1} is a harshad number')
    else:
        print(f'{num1} not a harshad number' )
    
def one_number(numlist):
    n = map(str, numlist)
    n= ''.join(n)
    n = int(n)
    return n

harshad_num()

Enter a number21
the number 21 is a harshad number


In [100]:
#99. write a function select_word(): to get a random word from a list of words use module random and call random.randint() to chose random
from random import randint

def select_word():
    word = [str(x) for x in input('Enter the words separated by comma').split(',')]
    choice = randint(0,len(word))
    return word[choice]

select_word()

Enter the words separated by commaapple,mango,fence,jug


'jug'

In [101]:
#100. write a function blank(word): where it should print _ after first letter to the range of length of word i.e apple= a _ _ _ _
def blank():
    word = input('enter the word')
    n= len(word)-1
    blank_ = word[0]+("_ ")*n
    return blank_

blank()

enter the wordrooster


'r_ _ _ _ _ _ '

In [102]:
from random_word import RandomWords
r = RandomWords()
clue = ""

def hang_man(word,clue,letter,try_left):
    # create a variable "try" and set it to True
    Try = True
    #--------------------------------
    # loop through the length of the word,
    for i in range(len(word)):
        # check condition if word is equal to letter
        #the set clue= clue from 0 to current index of loop + letter + clue from current index +1 to last
        if(word[i] == letter):
            clue = clue[:i] + letter + clue[i+1:]
            #set try to false
            Try = False
    #---------------------------------------------------
    # check condition if try is true then decrease the try_left by 1
    if(Try):
        try_left -= 1
        print("wrong try", try_left)
    print("you got so far",clue)
    
    # check if word is equal to clue:
    if(word == clue):
        print("you win")
    # check is try_left is 0:
    elif(not try_left):
        print("you lost, it was", word)
    else:
        # ask user to input next letter
        next_letter=input("enter the next letter: ")
        # call the recursive function of itself
        hang_man(word,clue,next_letter,try_left)

word = r.get_random_word()
clue = ''.join(["_" for i in range(0,len(word))])
hang_man(word,clue,word[0],5)

you got so far c_______
enter the next letter: h
wrong try 4
you got so far c_______
enter the next letter: a
you got so far ca_____a
enter the next letter: l
wrong try 3
you got so far ca_____a
enter the next letter: r
you got so far ca_r___a
enter the next letter: e
you got so far ca_re__a
enter the next letter: i
wrong try 2
you got so far ca_re__a
enter the next letter: s
wrong try 1
you got so far ca_re__a
enter the next letter: t
you got so far ca_retta
enter the next letter: p
wrong try 0
you got so far ca_retta
you lost, it was cabretta


In [ ]:
#!pip install random-word

In [1]:
#102. write a map function which calculate the list multiple of 5: eg: x == [ 2, 3, 4, 5, 6] y == [10, 15, 20, 25, 30]

x = [2,3,4,5,6]    
y = map(lambda x:x*5,x)
print(list(y))


[10, 15, 20, 25, 30]


In [2]:
#103. write a map function which multiply the odd values by 5 and add them to the list
x = [2,3,4,5,6,7,8]

y = map(lambda x:x*5 if (x%2!=0) else x == None,x)
print([item for item in list(y) if item!=False])

[15, 25, 35]


In [3]:
#104. write a map function to get: x=[2,3,4] y=[4,5] ans: z=[[6,7],[7,8],[8,9]]

x = [2,3,4]
y =[4,5]
z = [[x[i]+y[j] for j in range(len(y))] for i in range(len(x))]
print(z)

        

[[6, 7], [7, 8], [8, 9]]


In [4]:
#105 & 106. write a function to convert a z list to 1d list : use reduce
from functools import reduce
z_list = [[1, 2, 3],
            [4, 5, 6],
            [7, 8, 9]]

oned_list = reduce(lambda a,b:a+b,z_list)
print(oned_list)

[1, 2, 3, 4, 5, 6, 7, 8, 9]


In [25]:
#107. write a function using a type annotations which take a two vector a and b which calculate the vector sum eg: [1,2,3]+[4,5,6]=[5,7,9]
a = [1,2,3]
b= [4,5,6]
def add(a : list,b : list) ->list:
    c :list = map(sum,zip(a,b))
    print(list(c))

add(a,b)


[5, 7, 9]


In [26]:
#108. write a function which using a type annotation to calculate te dot product of two vector
a= [1,2,3]
b=[4,5,6]

def dotprd(a:list,b:list)->list:
    c:list = sum([x*y for x,y in zip(a,b)])
    print(c)

dotprd(a,b)

32


In [34]:
#109. write a function to calculate the sum of vector which take a list of vector using type annotation eg:[1,2],[3,4],[5,6]=[9,12]
from typing import List

x = [1,2]
y =[3,4]
z=[5,6]

def sumofvectors(lists : List[list]) ->list:
    res =[0]*len(lists[0])
    for l in lists:
        res = [item1+item2 for item1,item2 in zip(res,l)]
    return res

    
sumofvectors([x,y,z])


[9, 12]

In [30]:
#!pip install typing

In [39]:
#110. write a function to get the shape of a matrix also use type annotation: eg:`[[1,2],[2,3]]`=`(2,2)`
from typing import Tuple
Matrix =List[list[float]]

x = [[1,2],[2,3]]


def shapeOfMatrix(M:Matrix)->Tuple[int,int]:
    return (len(M),len(M[0]))

shapeOfMatrix(x)

(2, 2)

In [38]:
#111. write a functions to create a matrix and used function to crete a 3*3 identity matrix : hint:use a Callable from typing

from typing import Callable
# Creating type alias for a matrix
Matrix = List[List[float]]

def matrixer(num_rows: int, num_cols: int, entry_fn: Callable[[int, int], float]) -> Matrix:
    return [[entry_fn(i, j) for j in range(num_cols)] for i in range(num_rows)]

def identity_matrix(n: int) -> Matrix:
    return matrixer(n, n, lambda i, j: 1 if i == j else 0)

identity_matrix(3)



[[1, 0, 0], [0, 1, 0], [0, 0, 1]]

[callable from typing](https://www.humaneer.org/blog/creating-matrices-in-python-tutorial/)

In [41]:
#112. reber_ grammar

reber_grammar = [[("B",1)], [("T",2),("P",3),("S",2),("X",4),
                        ("T",3),("V",5),("X",3),("S",6),("P",4),("V",6),("E",None)]]

In [87]:
#113 write a function to generate a password :
#password should have a number,special_char,alphabet(both capital and small)
#allow user to input the length of password

import string
import random


def password():
    pwd_len = int(input('Enter the length of password: '))
    Bchar_len = int(0.35*pwd_len)
    Schar_len = int(0.25*pwd_len)
    numb_len = int(0.25*pwd_len)
    splchar_len = int(0.25*pwd_len)
    Bchar = [random.choice(string.ascii_uppercase) for i in range(Bchar_len) ]
    Schar = [random.choice(string.ascii_lowercase) for i in range(Schar_len) ]
    numb = [random.choice(string.digits) for i in range(numb_len) ]
    splchar = [random.choice(string.punctuation) for i in range(splchar_len) ]
    pwd = Bchar+Schar+numb+splchar   
    random.shuffle(pwd)
    return ''.join(pwd)
   

password()

    

Enter the length of password: 20


'V8jbq1}=K<0l4CP^IX>Xz0'

In [86]:
#114. write a function to get the integers satisfying a2+b2=c2 called Pythagorean numbers.
#the function take a input from a user with a maximal number: eg: if user give: 5 it should print (3,4) similarly:10 3 ,4 6,8

def pythagorean(lst,c):
    n = len(lst)
    for a in range(1,n):
        for b in range(a+1,n):
            if (lst[a]**2+lst[b]**2 == c**2):
                print('The values satisfying pythagorean to target',c,' is ',(lst[a],lst[b]))
                

c = int(input('Enter a maximal number'))
lst = list(range(1,(c**2)+1))
pythagorean(lst,c)


Enter a maximal number5
The values satisfying pythagorean to target 5  is  (3, 4)


115: write a program to guess a random number between range 1 to n

hint use random.randit() to generate a random number
the program should have upper bound and lower bound eg(1-30)
if user give 0 should exit the program
if the guess is less thean lower or grater than upper ("not in boundary)

In [120]:
def random_guess(n,guess,try_left):
            if guess==0:
                return 'Program exitted'
            elif guess>upper and guess<lower:
                print('Checking the boundry..')
                return 'not in boundary'
            else:
                if guess == n:
                    print('Number correct',guess,'\nThe computer generated: ',n)
                    
                elif(not try_left):
                    print('You lost. The number is ',n)
                
                else:
                    print('wrong, try again')
                    try_left-=1
                    print('try remaining',try_left)
                    guess = int(input('Enter a number again'))
                    random_guess(n,guess,try_left)



upper = int(input('Enter upper limit'))
lower = int(input('Enter lower limit'))
n=random.randint(lower,upper)
guess = int(input('Enter a number'))
random_guess(n,guess,4)
    

Enter upper limit30
Enter lower limit1
Enter a number5
wrong, try again
try remaining 3
Enter a number again7
wrong, try again
try remaining 2
Enter a number again3
wrong, try again
try remaining 1
Enter a number again6
wrong, try again
try remaining 0
Enter a number again8
You lost. The number is  4


116: Write a function which takes a text and encrypts it with a Caesar cipher as well as decrypt.(https://en.wikipedia.org/wiki/Caesar_cipher) .

input:"hello",
output: encrypt:EBIIL
decyprt: HELLO

In [147]:
def encrpyt(word,shift):    #ord returns the unicode of the character
    cipher=[]               #chr returns the character corresponding to the unicode. lower and uppercase letters have the same unicode
    for ch in word:
        n = ord(ch)
        n=n-shift
        if n>ord('z'):
            n=n+26
        cipher_letter = chr(n)
        cipher.append(cipher_letter)
    return ''.join(cipher)

def decrpyt(word,shift):
    decipher=[]
    for ch in word:
        n = ord(ch)
        n=n+shift
        if n>ord('z'):
            n=n+26
        decipher_letter = chr(n)
        decipher.append(decipher_letter)
    return ''.join(decipher)


message = str(input("Enter the message to be encrpyted"))
shift = int(input('Enter the shift value'))
encrpyted = encrpyt(message,shift)
decrpyted = decrpyt(encrpyted,shift)
print('The encrpyted message: ',encrpyted)
print('The decrpyted message: ',decrpyted)



Enter the message to be encrpytedHello
Enter the shift value3
The encrpyted message:  Ebiil
The decrpyted message:  Hello


In [ ]:
new_dict = dict([(value, key) for key, value in old_dict.items()])

In [95]:
#118: Write a function which translates a text to morse code
morse_dict = {'A':'.-', 'B':'-...', 'C':'-.-.', 'D':'-..', 
      'E':'.', 'F':'..-.', 'G':'--.','H':'....', 'I':'..','J':'.---','K':'-.-','L':'.-..','M':'--','N':'-.','O':'---',
      'P':'.--.','Q':'--.-','R':'.-.','S':'...','T':'-','U':'..-','V':'...-','W':'.--','X':'-..-','Y':'-.--','Z':'--..',
      1:'.----',2:'..---',3:'...--',4:'....-',5:'.....',6:'-....',7:'--...',8:'---..',9:'----.',0:'-----'}

string_dict = dict([(value,key) for key,value in morse_dict.items()])

def morse_code(morse_dict,text):
    text=text.upper()
    return ' '.join([morse_dict[ch] for ch in text])

def message(string_dict,code):
    code = code.split()
    return ''.join([string_dict[chr] for chr in code])

text = 'Python'
enc = morse_code(morse_dict,text)
dec= message(string_dict,enc)
print(text,'-->',enc,'-->',dec)

Python --> .--. -.-- - .... --- -. --> PYTHON


In [151]:
#118. Write a program to calculate the square root of a number by using the Babylonian method.(https://en.wikipedia.org/wiki/Methods_of_computing_square_roots)
number = abs(float(input('Enter the number to find the squareroot: ')))
guess = abs(float(input('Enter the first guess')))
epsilon = 0.001
print('Difference : Squareroot')
while(True):
    difference = guess**2 - number
    print('{} : {}'.format(round(difference,4),round(guess,4)))
    if abs(difference)<=epsilon:
        break
    guess = (guess + number/guess)/2.0
    
    


Enter the number to find the squareroot: 125678
Enter the first guess400
Difference : Squareroot
34322.0 : 400.0
1840.6245 : 357.0975
6.642 : 354.5203
0.0001 : 354.5109


The Babylonian Method states that if the previous guess, xn, is an overestimate of the square root of a number, S, then a more precise next guess, xn+1, is the average of the previous guess and the number divided by the previous guess.

xn+1 = ( xn + S / xn ) / 2

In [164]:
def multiple(N,n):
    if n>10:
        return
    else:
        print('Multiple of ',n,'=', N*n)
        return multiple(N,n+1)
N=int(input('Enter the number: '))
multiple(N,1)

Enter the number: 7
Multiple of  1 = 7
Multiple of  2 = 14
Multiple of  3 = 21
Multiple of  4 = 28
Multiple of  5 = 35
Multiple of  6 = 42
Multiple of  7 = 49
Multiple of  8 = 56
Multiple of  9 = 63
Multiple of  10 = 70


In [206]:
#121. Write a function which implements the Pascal's triangle use recursion

from math import factorial

def pascal(N):
    for i in range(N):
        for j in range(i+1):
            k = factorial(i)/(factorial(j)*factorial(i-j))
            print(' ',int(k),end='')
        print('\n')

N = int(input("Enter the number of rows in pascal's traingle: "))
pascal(N)
    

    

Enter the number of rows in pascal's traingle: 10
  1

  1  1

  1  2  1

  1  3  3  1

  1  4  6  4  1

  1  5  10  10  5  1

  1  6  15  20  15  6  1

  1  7  21  35  35  21  7  1

  1  8  28  56  70  56  28  8  1

  1  9  36  84  126  126  84  36  9  1



In [202]:
import sys
sys.setrecursionlimit(1500)

In [208]:
#123: write a fuction to check if the given data type are iterable or not ?
def checker(target):
    try:
        iter(target)
    except TypeError:
        print('Data type not iterable')
    else:
        print('Data type is iterable')

var1 = [1,2,3,4]
var2 =(2,5,'var',7)
var3 = 'Checking'
var4 = 5
print('var1 :')
checker(var1)
print('var2 :')
checker(var2)
print('var3 :')
checker(var3)
print('var4 :')
checker(var4)


var1 :
Data type is iterable
var2 :
Data type is iterable
var3 :
Data type is iterable
var4 :
Data type not iterable


124: Write a generator which computes the running average.

    eg:[1,2,3,4,5,6]
        1-->1
        2---->1.50
        3----->2
        4----->2.50
        5---->3
        6---->3.50

In [24]:
import numpy as np

def moving_average(x,N):
    yield np.convolve(x, np.ones(N)/N, mode='valid')
    
arr = [1,2,3,4,5,6]
window=2
moving_average(arr,window)

<generator object moving_average at 0x7f8c24316c10>

[moving average in python](https://stackoverflow.com/questions/13728392/moving-average-or-running-mean)

In [25]:
list(moving_average(arr,window))

[array([1.5, 2.5, 3.5, 4.5, 5.5])]

127: Write a generator cycle performing the same task.

- cycle('ABCD') --> A B C D A B C D A B C D ...

In [33]:
def cycle(word):
    count=1
    while True:
        yield word*count
        count+=1

In [39]:
gen = cycle('ABCD')

'ABCD'

In [46]:
next(gen)

'ABCDABCDABCDABCDABCDABCDABCDABCD'

126:Write a generator which returns  a zero or a one in every iteration. 
- The probability p for returning a 1 is defined in a variable p.
- default value of probability=0.5. In other words,1 and 0 will be returned with the same probability.

In [52]:
import random
def proba(data,probability):
    count=1
    while True:
     num = random.choices(data,probability)
     yield num
     count+=1


gen2= proba([0,1],[0.5,0.5])

In [56]:
#first generation
next(gen2)

[0]

In [57]:
#second generation
next(gen2)

[1]

In [58]:
#third generation
next(gen2)

[0]

In [59]:
next(gen2)

[1]

In [68]:
#130: write a nested function to calulate the temp from celcius to farhen
def outer(C):
    def inner():
        f = (C*1.8)+32
        print(f'Temp in Farahenheit : %.2f' %f)
    inner()

C = float(input('Enter the temperature in celcius'))
outer(C)

Temp in Farahenheit : 89.60


131:Write a nested  function so that it can work for polynomials of arbitrary degree:

\begin{equation}
\sum_{k=0}^{n} a_{k} \cdot x^{k}=a_{n} \cdot x^{n}+a_{n-1} \cdot x^{n-1}+\ldots+a_{2} \cdot x^{2}+a_{1} \cdot x+a_{0}
\end{equation}

In [71]:
def equation(a,x):
    def calculation(const,x,k):
        return const*(x**k)
    summation=0
    for k in range(len(a)):
        summation +=calculation(a[k],x,k)   
    return summation

equation([1,2,3,4],2)     

49

In [74]:
#132:write a decorator so that the argument passed is. positive integer for a function factorial that you have created before
def positive_checker(num):
    def wrapper(num):
        if num>0:
            return num
        else:
            print('Enter a positive number')
    return wrapper
    
@positive_checker        
def recursive(num):
    if num==1:
        return num
    else:
        return num*recursive(num-1)

num = int(input('Enter a number: '))
print('The factorial is ',recursive(num))


Enter a positive number
The factorial is  None


In [87]:
#133:write a decorator to count the number of times a function has beeen called . eg
#define a add function and loop it for 10 time and check

def call_checker(f):
    def inner_wrap(*args,**kwargs):
        inner_wrap.count+=1
        return f(*args,**kwargs)
    inner_wrap.count=0
    return inner_wrap()

@call_checker
def call_function():
    for i in range(10):
        print('round ',i)
        
try:
    call_function()
except TypeError:
    print('Ignored.counted successfully')        

round  0
round  1
round  2
round  3
round  4
round  5
round  6
round  7
round  8
round  9
counted successfully
